tools/generator/src/Extensions/XCScheme+Extensions.swift

```diff
@@ -0,0 +1,126 @@
+import XcodeProj
+
+enum XCSchemeConstants {
+    // GH399: Derive the defaultLastUpgradeVersion
+    static let defaultLastUpgradeVersion = "1320"
+    static let lldbInitVersion = "1.7"
+}
+
+// MARK: XCScheme.PBXTargetInfo
+
+extension XCScheme {
+    struct PBXTargetInfo {
+        let pbxTarget: PBXTarget
+        let buildableReference: XCScheme.BuildableReference
+
+        init(pbxTarget: PBXTarget, referencedContainer: String) {
+             self.pbxTarget = pbxTarget
+             buildableReference = .init(
+                 pbxTarget: pbxTarget,
+                 referencedContainer: referencedContainer
+             )
+        }
+    }
+}
+
+// MARK: XCScheme.BuildableReference
+
+extension XCScheme.BuildableReference {
+    convenience init(pbxTarget: PBXTarget, referencedContainer: String) {
+        self.init(
+            referencedContainer: referencedContainer,
+            blueprint: pbxTarget,
+            buildableName: pbxTarget.buildableName,
+            blueprintName: pbxTarget.name
+        )
+    }
+}
+
+// MARK: XCScheme.BuildAction
+
+extension XCScheme.BuildAction {
+    convenience init<PBXTargetInfos: Sequence>(
+        buildMode: BuildMode,
+        targetInfos: PBXTargetInfos,
+        hostBuildableReference: XCScheme.BuildableReference?,
```

Comment:
For custom schemes there could be multiple hosts, right?

Reply:
Perhaps, I am not thinking about this correctly. In the current code, we [generate a scheme per host](https://github.com/buildbuddy-io/rules_xcodeproj/blob/main/tools/generator/src/Generator/CreateAutogeneratedXCSchemes.swift#L103-L114). I was thinking that we would need to do something similar for custom schemes. 

Alternatively, should we be selecting the "best" host from `hostPBXTargets`?

Reply:
Host + hosted target constitute a single buildable scheme target. Right now all of our schemes only have one buildable target. With custom schemes we will be able to have multiple buildable targets. If one of those custom schemes includes two watch apps, then it would have two `hostBuildableReference` as well.

Reply:
Looking at this again, I think that I should associate the hosts with the `PBXTargetInfo`.  Then, I can add each host as a build action entry. I will update the PR.